### PR TITLE
【Hackathon No.10】AdaptiveLogSoftmaxWithLoss API

### PR DESCRIPTION
**Pull request by gsq7474741**
### PR types

New features
### PR changes

APIs
### Describe

在 Paddle 框架中，新增 AdaptiveLogSoftmaxWithLoss API
RFC：[20200322_api_design_for_AdaptiveLogSoftmaxWithLoss.md](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20200322_api_design_for_AdaptiveLogSoftmaxWithLoss.md)
文档PR：https://github.com/PaddlePaddle/docs/pull/4650

### CODE REVIEW

**Comment by Ligoml** [comment]
需要在Describe中添加rfc链接和中文文档pr链接，待CI大部分通过后，开启review

**Comment by gsq7474741** [comment]
> 需要在Describe中添加rfc链接和中文文档pr链接，待CI大部分通过后，开启review

已添加

**Comment by gsq7474741** [comment]
@Ligoml ci也基本都过了

**Comment by gsq7474741** [comment]
@yingyibiao 已更新~

**Comment by yingyibiao** [comment]
需要新增 `paddle.nn.functional.adaptive_log_softmax_with_loss` API

**Comment by gsq7474741** [comment]
> 需要新增 `paddle.nn.functional.adaptive_log_softmax_with_loss` API

这个api里面需要构造一些对象，且需要保留，所以我觉得做成Layer比较合适，本任务提出的functional api可能不适用，不知您怎么看

File: python/paddle/nn/layer/loss.py [hunk 1]
```diff
@@ -15,14 +15,279 @@
...
 # TODO: define loss functions of neural network
 import numpy as np
+from collections import namedtuple
 import paddle.fluid as fluid
 import paddle
 from .. import functional as F
+from .. import Layer, Sequential, LayerList
+from paddle.nn.layer import Linear
 from paddle.fluid.framework import _varbase_creator
 from .. import Layer
```
Comment (yingyibiao) [inline python/paddle/nn/layer/loss.py:25]: Layer 重复导入

Comment (yingyibiao) [reply python/paddle/nn/layer/loss.py:25]: import 包的顺序按照 built-in、third-party、项目内部包这三类区分。

**Comment by yingyibiao** [comment]
> > 需要新增 `paddle.nn.functional.adaptive_log_softmax_with_loss` API
> 
> 这个api里面需要构造一些对象，且需要保留，所以我觉得做成Layer比较合适，本任务提出的functional api可能不适用，不知您怎么看

可以把具体的理由详细说明一下吗？adaptive_log_softmax_with_loss 和 paddle.nn.functional 目录下其他 loss 函数的差异点在哪里呢？

File: python/paddle/nn/layer/loss.py [hunk 2]
```diff
@@ -15,14 +15,279 @@
...
+        elif not_in_shortlist.all():
+            log_prob = self._get_full_log_prob(input, head_output)
+            return paddle.argmax(log_prob, axis=1)
+
+        else:
+            log_prob = self._get_full_log_prob(input[not_in_shortlist],
+                                               head_output[not_in_shortlist])
+            output[not_in_shortlist] = paddle.argmax(
+                log_prob, axis=1).cast('float32')
+            return output
```
Comment (yingyibiao) [inline python/paddle/nn/layer/loss.py:290]: AdaptiveLogSoftmaxWithLoss 这个类的代码&注释和 torch 一摸一样，原则上我们是禁止直接搬运 torch 代码的。

Comment (gsq7474741) [reply python/paddle/nn/layer/loss.py:290]: 注释里加了pytorch的license，不知是否可以，而且这个代码部分适配paddle做了一点修改

Comment (yingyibiao) [reply python/paddle/nn/layer/loss.py:290]: 可以借鉴 pytorch 代码，但是不能一大段直接搬运。

Comment (gsq7474741) [reply python/paddle/nn/layer/loss.py:290]: 好的 我改一下
